### PR TITLE
Bus cleanup patch

### PR DESCRIPTION
**Pull request by Fatsie**
This is a clean-up patch to prepare for the TypedBus patch as discussed in #1186.
Some extra comments next to commit messages:
* I did delete AD9361 driver as it seemed to be wrongly subclassing ``BusDriver`` and I don't think this code is maintained. Alternatively I could have it subclass ``Driver`` but I have no way of checking that patch.
* I removed ``AvalonSTPktsWithChannel`` and added ``report_channel`` parameter to ``AvalanSTPkts.__init__()`` which I think should allow for the same functionality. I need input if that is not the case.
  It does mean that code using ``AvalonSTPktsWithChannel`` has to be adapted.

### CODE REVIEW

File: cocotb/drivers/ad9361.py [hunk 1]
```diff
@@ -1,214 +0,0 @@
...
-
-import cocotb
-from cocotb.triggers import Timer, RisingEdge, Event
-from cocotb.drivers import BusDriver
-from cocotb.binary import BinaryValue, BinaryRepresentation
-
-from collections import deque
-
-
-class AD9361(BusDriver):
```
Comment (eric-wieser) [inline cocotb/drivers/ad9361.py]: Instead of deleting this class as part of this Pr, I'd just remove the base class, and open an issue about removing it.

File: cocotb/drivers/__init__.py [hunk 2]
```diff
@@ -221,26 +220,22 @@ class BusDriver(Driver):
...
     
     _optional_signals = []
 
     def __init__(self, entity, name, clock, **kwargs):
-        # emulate keyword-only arguments in python 2
-        index = kwargs.pop("array_idx", None)
-        bus_separator = kwargs.pop("bus_separator", "_")
-        reject_remaining_kwargs('__init__', kwargs)
+        index = kwargs.get("array_idx")
+        kwargs["optional_signals"] = self._optional_signals + kwargs.get("optional_signals", [])
```
Comment (eric-wieser) [inline cocotb/drivers/__init__.py]: It's not clear to me why this makes sense - what is the use case of passing extra optional signals here?

File: cocotb/drivers/avalon.py [hunk 3]
```diff
@@ -512,9 +513,8 @@ class AvalonST(ValidatedBusDriver):
 
     _default_config = {"firstSymbolInHighOrderBits" : True}
 
-    def __init__(self, *args, **kwargs):
-        config = kwargs.pop('config', {})
-        ValidatedBusDriver.__init__(self, *args, **kwargs)
+    def __init__(self, entity, name, clock, *, config={}, **kwargs):
```
Comment (eric-wieser) [inline cocotb/drivers/avalon.py]: This is python3-only syntax, and we haven't yet decided what to do for #767. At any rate, we should probably aim to do at least one more python 2 release.

File: cocotb/drivers/opb.py [hunk 4]
```diff
@@ -47,8 +47,8 @@ class OPBMaster(BusDriver):
     _optional_signals = ["seqAddr"]
     _max_cycles = 16
 
-    def __init__(self, entity, name, clock):
-        BusDriver.__init__(self, entity, name, clock)
+    def __init__(self, entity, name, clock, **kwargs):
+        BusDriver.__init__(self, entity, name, clock, **kwargs)
```
Comment (eric-wieser) [inline cocotb/drivers/opb.py]: All these `BusDriver.__init__` changes are great, thanks

File: cocotb/monitors/__init__.py [hunk 5]
```diff
@@ -174,14 +174,16 @@ class BusMonitor(Monitor):
...
+                 **kwargs):
         self.log = SimLog("cocotb.%s.%s" % (entity._name, name))
         self.entity = entity
         self.name = name
         self.clock = clock
         self.bus = Bus(self.entity, self.name, self._signals,
                        optional_signals=self._optional_signals,
-                       bus_separator=bus_separator, array_idx=array_idx)
+                       bus_separator=bus_separator, array_idx=array_idx,
+                       **kwargs)
```
Comment (eric-wieser) [inline cocotb/monitors/__init__.py]: I'm not sold on this - what kwargs are you expecting?

File: cocotb/drivers/ad9361.py (hunk 1 above)
Comment (Fatsie) [reply cocotb/drivers/ad9361.py]: I'll leave the AS9361 in as is then and make separate issue for removing it.
The code will only cause problems when actually trying to create an object of that type and also when removing the base class.

File: cocotb/monitors/__init__.py (hunk 5 above)
Comment (Fatsie) [reply cocotb/monitors/__init__.py]: In follow-up patch introducing ``TypedBus`` the Bus class will not be fixed and different call signatures need to be supported. But I can do that in the follow-up patch then. I will also plan to rename kwargs to bus_kwargs then. I can either move this patch to follow-up patch or do the following:
```python
class BusMonitor(Monitor):
    """Wrapper providing common functionality for monitoring buses."""
    _signals = []
    _optional_signals = []

    def __init__(self, entity, name, clock, reset=None, reset_n=None,
                 callback=None, event=None, **bus_kwargs):
        self.log = SimLog("cocotb.%s.%s" % (entity._name, name))
        self.entity = entity
        self.name = name
        self.clock = clock
        self.bus = Bus(self.entity, self.name, self._signals,
                       optional_signals=self._optional_signals,
                       **bus_kwargs)
```

File: cocotb/drivers/avalon.py (hunk 3 above)
Comment (Fatsie) [reply cocotb/drivers/avalon.py]: So the question is then what is preferred:
```python
    def __init__(self, entity, name, clock, config={}, **kwargs):
```
or
```python
    def __init__(self, entity, name, clock, **kwargs):
        config = kwargs.pop('config', {})
```
I like the former because then the parameter is shown when doing help in ipython etc. but then people could wrongly use positional passing of config.

Comment (eric-wieser) [reply cocotb/drivers/avalon.py]: I would go for the latter - we can fix the docs as soon as we drop python 2, but we can't change the API once we release it.

File: cocotb/monitors/__init__.py (hunk 5 above)
Comment (eric-wieser) [reply cocotb/monitors/__init__.py]: Let's move this to the follow-up, I think it will be easier to understand in context.

File: cocotb/drivers/ad9361.py (hunk 1 above)
Comment (eric-wieser) [reply cocotb/drivers/ad9361.py]: Does this even use a single method from `BusDriver`? If not, removing the base class ought to be perfectly safe

File: cocotb/monitors/__init__.py (hunk 5 above)
Comment (Fatsie) [reply cocotb/monitors/__init__.py]: I'll move it to follow-up patch, it's more natural there.

File: cocotb/drivers/avalon.py (hunk 3 above)
Comment (Fatsie) [reply cocotb/drivers/avalon.py]: Will keep use of pop for the moment; can be reconsidered later.

**Comment by Fatsie** [comment]
* Fully reworked patch based on Erik's comments.
* Some extra enforcements of passing by name using ``kwargs.pop()``

File: cocotb/drivers/amba.py [hunk 6]
```diff
@@ -236,10 +236,11 @@ class AXI4Slave(BusDriver):
...
     ]
 
-    def __init__(self, entity, name, clock, memory, callback=None, event=None,
-                 big_endian=False):
-
-        BusDriver.__init__(self, entity, name, clock)
+    def __init__(self, entity, name, clock, memory, **kwargs):
+        callback = kwargs.pop('callback', None)
+        event = kwargs.pop('event', None)
+        big_endian = kwargs.pop('big_endian', False)
```
Comment (eric-wieser) [inline cocotb/drivers/amba.py]: Making these keyword-only when they used not to be is a breaking change, sadly. I don't think we should risk breaking existing users for such minor cleanup.

Comment (Fatsie) [reply cocotb/drivers/amba.py]: I'll move it to follow-up patch.

**Comment by Fatsie** [comment]
Only change is removing patch enforcing passing default parameters by name.
Moved to follow-up patch.

File: cocotb/drivers/__init__.py [hunk 7]
```diff
@@ -215,32 +214,29 @@ class BusDriver(Driver):
...
-            bus_separator (str, optional): Character(s) to use as separator between bus
-                name and signal name. Defaults to '_'.
+    Args:
+        entity (SimHandle): A handle to the simulator entity.
+        name (str or None): Name of this bus. ``None`` for a nameless bus, e.g.
+            bus-signals in an interface or a ``modport``.
+            (untested on ``struct``/``record``, but could work here as well).
+        clock (SimHandle): A handle to the clock associated with this bus.
+        **kwargs (dict): kwargs is passed on to __init__ of the created .bus field.
+            See ``cocotb.Bus.__init__()`` for more information.
```
Comment (eric-wieser) [inline cocotb/drivers/__init__.py]: Perhaps just ``Keyword arguments forwarded to :class:`cocotb.Bus`, see the docs for that class for more information.``

File: cocotb/monitors/avalon.py [hunk 8]
```diff
@@ -100,15 +100,20 @@ class AvalonSTPkts(BusMonitor):
...
 
         self.config = self._default_config.copy()
+        self.report_channel = report_channel
 
         # Set default config maxChannel to max value on channel bus
         if hasattr(self.bus, 'channel'):
             self.config['maxChannel'] = (2 ** len(self.bus.channel)) -1
+        else:
+            if report_channel:
+                self.log.error("Channel reporting asked on bus without channel signal")
```
Comment (eric-wieser) [inline cocotb/monitors/avalon.py]: This should be `raise ValueError(...)`

File: cocotb/monitors/avalon.py [hunk 9]
```diff
@@ -222,20 +230,3 @@ def valid():
...
-    def __init__(self, *args, **kwargs):
-        AvalonSTPkts.__init__(self, *args, **kwargs)
-
-    def _recv(self, pkt):
-        """Force use of channel in recv function.
-
-        Args:
-            pkt: (string) Monitored data.
-        """
-        AvalonSTPkts._recv(self, {"data": pkt, "channel": self.channel})
```
Comment (eric-wieser) [inline cocotb/monitors/avalon.py]: Again, removing this class will break any existing code that uses it.

It looks like you could keep old code working by removing everything but `__init__`, and changing it to pass `record_channel=True` to `__init__`? Perhaps:
```
def __init__(*args, **kwargs):
    warnings.warn(
        "AvalonSTPktsWithChannel is deprecated, use AvalonSTPkts(..., record_channel=True)",
        DeprecationWarning, stacklevel=2)
    AvalonSTPkts.__init__(self, *args, record_channel=True, **kwargs)

File: cocotb/monitors/avalon.py [hunk 10]
```diff
@@ -100,15 +100,20 @@ class AvalonSTPkts(BusMonitor):
         "invalidTimeout"                : 0,
     }
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, entity, name, clock, **kwargs):
         config = kwargs.pop('config', {})
-        BusMonitor.__init__(self, *args, **kwargs)
+        report_channel = kwargs.pop('report_channel', False)
```
Comment (eric-wieser) [inline cocotb/monitors/avalon.py]: Please document this new argument

File: cocotb/monitors/avalon.py (hunk 8 above)
Comment (Fatsie) [reply cocotb/monitors/avalon.py]: OK, that was original version of patch but changed it to use self.log.error. Any guidance on when the raise and when to use self.log.error ?

Comment (eric-wieser) [reply cocotb/monitors/avalon.py]: Use `raise` if the caller called your function in an invalid way, and it doesn't make sense to continue.

Use `self.log.error` if the bus itself is misbehaving, and throwing an error immediately would leave the bus in an invalid state. Even if you do eventually throw an exception, if you weren't able to do it immediately, you should also `log.error` so that the sim time of when things went wrong is recorded.

TL;DR: `log.error` is for humans only, `raise` is for both humans and code.

File: cocotb/monitors/avalon.py (hunk 9 above)
Comment (Fatsie) [reply cocotb/monitors/avalon.py]: Can I use [deprecated](https://pypi.org/project/Deprecated/) ? I really think we should get rid of this old legacy somewhere in the future.

Comment (eric-wieser) [reply cocotb/monitors/avalon.py]: I'd prefer not to in this patch. Cocotb currently has no dependencies, we should have that discussion independently.

Perhaps open an issue suggesting we replace all instances of `DeprecationWarning` with that third party module, and keep doing it by hand for now.

Agreed that it's nice to remove legacy, but if it's user facing we should always deprecate first.

Comment (Fatsie) [reply cocotb/monitors/avalon.py]: warn & DeprecationWarning is fine. Was just used to @deprecated.

File: cocotb/drivers/__init__.py (hunk 7 above)
Comment (Fatsie) [reply cocotb/drivers/__init__.py]: OK

**Comment by Fatsie** [comment]
Handled Erik's comments

File: cocotb/monitors/avalon.py [hunk 11]
```diff
@@ -224,18 +241,15 @@ def valid():
...
+    """
 
-        Args:
-            pkt: (string) Monitored data.
-        """
-        AvalonSTPkts._recv(self, {"data": pkt, "channel": self.channel})
+    def __init__(self, entity, name, clock, **kwargs):
+        warnings.warn(
+            "Use of AvalonSTPktsWithChannel is deprecated\n"
+            "\tUse AvanlonSTPkts(..., report_channel=True, ...)",
```
Comment (eric-wieser) [inline cocotb/monitors/avalon.py]:
```suggestion
            "\tUse AvalonSTPkts(..., report_channel=True, ...)",
```

File: cocotb/monitors/avalon.py [hunk 12]
```diff
@@ -224,18 +241,15 @@ def valid():
...
+            DeprecationWarning
```
Comment (eric-wieser) [inline cocotb/monitors/avalon.py]:
```suggestion
            DeprecationWarning, stacklevel=2
```

File: cocotb/monitors/avalon.py (hunk 8 above)
Comment (cmarqu) [reply cocotb/monitors/avalon.py]: > TL;DR: `log.error` is for humans only, `raise` is for both humans and code.

This text should be part of a "How to write cocotb code", maybe in #1141.

**Comment by Fatsie** [comment]
Squashed corrections.

File: cocotb/drivers/__init__.py [hunk 13]
```diff
@@ -215,32 +214,29 @@ class BusDriver(Driver):
...
+        index = kwargs.get("array_idx", None)
 
         self.log = SimLog("cocotb.%s.%s" % (entity._name, name))
         Driver.__init__(self)
         self.entity = entity
         self.clock = clock
         self.bus = Bus(self.entity, name, self._signals,
-                       self._optional_signals, bus_separator=bus_separator, array_idx=index)
+                       optional_signals=self._optional_signals, **kwargs
+        )
```
Comment (eric-wieser) [inline cocotb/drivers/__init__.py]: nit: this is some weird line wrapping. I'd use either:
```
x = foo(arg1,
        arg2)
```
or
```
x = foo(
    arg1,
    arg2
)

Comment (eric-wieser) [reply cocotb/drivers/__init__.py]: In fact, this is exactly what `flake8` is complaining about in CI:
```
./cocotb/drivers/__init__.py:239:9: E124 closing bracket does not match visual indentation
```

Comment (Fatsie) [reply cocotb/drivers/__init__.py]: Pushed change.

**Comment by Fatsie** [comment]
See #1194; DeprecationWarning not working for me.

**Comment by Fatsie** [comment]
Pushed following additional changes:
```diff
diff --git a/cocotb/monitors/avalon.py b/cocotb/monitors/avalon.py
index 1fa3645..405a43c 100644
--- a/cocotb/monitors/avalon.py
+++ b/cocotb/monitors/avalon.py
@@ -139,9 +139,6 @@ class AvalonSTPkts(BusMonitor):
         self.config["useEmpty"] = (num_data_symbols > 1)
 
         if hasattr(self.bus, 'channel'):
-            if "channel" in self._optional_signals:
-                self.log.warning("Channel is not fully implemented in this monitor. Recommend use of AvalonSTPktsWithChannel.")
-
             if len(self.bus.channel) > 128:
                 raise AttributeError("AvalonST interface specification defines channel width as 1-128. "
                                      "%d channel width is %d" %
```

File: cocotb/monitors/avalon.py [hunk 14]
```diff
@@ -210,7 +221,10 @@ def valid():
                     self.log.info("Received a packet of %d bytes", len(pkt))
                     self.log.debug(hexdump(str((pkt))))
                     self.channel = channel
-                    self._recv(pkt)
+                    if channel is not None and self.report_channel:
```
Comment (eric-wieser) [inline cocotb/monitors/avalon.py]:
```suggestion
                    if self.report_channel:
```
`channel` can never be `None` here anyway

**Comment by imphil** [comment]
Thanks for this work @Fatsie! I leave the content review to @eric-wieser, but please fix the commit messages to follow the template:

```
summary line (up to 50 characters)

longer description, wrap at 72 characters
```

Read more about that in https://github.com/cocotb/cocotb/blob/master/CONTRIBUTING.md#patch-requirements

**Comment by eric-wieser** [comment]
Commit message suggestions:

* add to the first one a comment explaining that the motivation is to keep the list of optional and required signals fixed
* Change the second one to something like
```
Use more explicit signatures the avalon driver

Replace `*args` in signatures with explicit argument names for clarity
```

**Comment by Fatsie** [comment]
Rebased; handled requested changes.

> summary line (up to 50 characters)
> 
> longer description, wrap at 72 characters

OK with wrapping at 72 character but I do feel 50 characters is reducing usefulness of summary. 10 more would be much better.
I complied though.

**Comment by Fatsie** [comment]
Looking at recent merges more than 50% seem to violate the 50 characters summary rule.

**Comment by imphil** [comment]
Thanks for your update. This PR is not the best place to discuss agreed-on policy, so thanks for opening a new issue.

In your commit messages here, you can save characters by not trying to format commit message in rst; they are plaintext.
